File: python/monte_carlo.py
from multiprocessing import Pool

# Linear Congruential Generator - same formula across all languages
class LCG:
    def __init__(self, seed):
        self.seed = seed
    
    def random(self):
        self.seed = (self.seed * 1664525 + 1013904223) & 0xFFFFFFFF
        return (self.seed & 0x7FFFFFFF) / 0x7FFFFFFF

def monte_carlo_worker(args):
    samples, seed = args
    rng = LCG(seed)
    inside = 0
    
    for _ in range(samples):
        x = rng.random()
        y = rng.random()
        if x*x + y*y <= 1.0:
            inside += 1
    
    return inside

def monte_carlo_operation(total_samples, num_workers):
    samples_per_worker = total_samples // num_workers
    remainder = total_samples % num_workers
    
    tasks = []
    for i in range(num_workers):
        samples = samples_per_worker
        if i == num_workers - 1:
            samples += remainder
        tasks.append((samples, 12345 + i * 67890))  # Consistent seed pattern
    
    with Pool(processes=num_workers) as pool:
        results = pool.map(monte_carlo_worker, tasks)
    
    total_inside = sum(results)
    pi_estimate = 4.0 * total_inside / total_samples
    
    print(f"Monte Carlo Pi Estimation")
    print(f"Total samples: {total_samples}")
    print(f"Points inside circle: {total_inside}")
    print(f"Pi estimate: {pi_estimate:.6f}")
    print(f"Error: {3.141592653589793 - pi_estimate:.6f}")

File: go/monte_carlo.go
package main

import (
	"fmt"
	"sync"
)

// Linear Congruential Generator - same formula across all languages
func lcgRandom(seed *uint32) float64 {
	*seed = *seed*1664525 + 1013904223
	return float64(*seed&0x7FFFFFFF) / float64(0x7FFFFFFF)
}

func monteCarloWorker(samples int, seed uint32) int {
	inside := 0

	for range samples {
		x := lcgRandom(&seed)
		y := lcgRandom(&seed)
		if x*x + y*y <= 1.0 {
			inside++
		}
	}

	return inside
}

func monteCarloOperation(totalSamples int, numWorkers int) {
	if numWorkers <= 0 {
		numWorkers = 1
	}

	samplesPerWorker := totalSamples / numWorkers
	remainder := totalSamples % numWorkers

	var wg sync.WaitGroup
	results := make(chan int, numWorkers)

	for i := range numWorkers {
		samples := samplesPerWorker
		if i == numWorkers-1 {
			samples += remainder
		}

		wg.Add(1)
		go func(workerID int, numSamples int) {
			defer wg.Done()
			seed := uint32(12345 + workerID*67890) // Consistent seed pattern
			inside := monteCarloWorker(numSamples, seed)
			results <- inside
		}(i, samples)
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	totalInside := 0
	for inside := range results {
		totalInside += inside
	}

	piEstimate := 4.0 * float64(totalInside) / float64(totalSamples)

	fmt.Printf("Monte Carlo Pi Estimation\n")
	fmt.Printf("Total samples: %d\n", totalSamples)
	fmt.Printf("Points inside circle: %d\n", totalInside)
	fmt.Printf("Pi estimate: %.6f\n", piEstimate)
	fmt.Printf("Error: %.6f\n", 3.141592653589793 - piEstimate)
}
